### PR TITLE
Indirect encoding support

### PR DESCRIPTION
**Pull request by sevaa**
Adds support for strx, addrx, loclistx, rnglistx forms, also for RLE_...x entries in rangelists, LLE_...x entries in loclists.

One kink of note: generally, resolving indirect-form attributes is possible at parse time, but with one exception: sometimes one would get an indirect-form attribute in the very top DIE of a CU where the needed base_addr attribute is, and before the respective base_addr. This requires either some kind of look-ahead logic, or a second pass. I went with the latter, and only for the top DIE.

### CODE REVIEW

File: elftools/dwarf/ranges.py [hunk 1]
```diff
@@ -16,19 +16,22 @@
 
 RangeEntry = namedtuple('RangeEntry', 'entry_offset entry_length begin_offset end_offset is_absolute')
 BaseAddressEntry = namedtuple('BaseAddressEntry', 'entry_offset base_address')
+# If we ever see a list with a base entry at the end, there will be an error that entry_length is not a field.
 
-def not_implemented(e):
-    raise NotImplementedError("Range list entry %s is not supported yet" % (e.entry_type,))
+def _translate_startx_length(e, cu):
+    # Avoid retrieving the start offset twice - I don't think Python can do inline assignments
```
Comment (eliben) [inline elftools/dwarf/ranges.py:22]: This comment can be removed, IMHO. It's fairly clear what's going on

**Review by eliben — COMMENTED** [review]
Nice!

One question: does the double parsing always happen now for the top CU or just when these attrs are present?

**Comment by sevaa** [comment]
The second round of parsing (`_translate_indirect_attributes()`) is always invoked for the top DIE, but *only* for the top DIE. It bypasses all attributes with forms that are not indirect and translates all ones that are. See compileunit.py:100 and die.py.320.

With some effort, I can make the second pass conditional upon the presence of attributes that can't be translated on the first pass, but this will incur a penalty for non-top DIEs. The challenge is that translating the indirect forms involves, in most straightforward implementation, calling `get_top_DIE()` to get the base address, and if the DIE being parsed *is* the top DIE, then you get an infinite recursion. Adding a check there will marginally slow things down for every other DIE.

File: elftools/dwarf/ranges.py (hunk 1 above)
Comment (sevaa) [reply elftools/dwarf/ranges.py]: Removed.

**Comment by sevaa** [comment]
Fixed.

**Review by eliben — APPROVED** [review]

**Comment by sevaa** [comment]
Okay. Now, I'll be AFK for a couple of weeks, but then, I have plans to revise my test codebase to reduce the size and build complexity while keeping the coverage. The thing with nontrivial DWARF operations is that they only come up when the code is both convoluted and optimized; you won't get them with "Hello world" type programs.

**Comment by eliben** [comment]
SGTM, I'll be AFK for a bit too. Enjoy your vacation and thanks for all the contributions recently!

When we've both back, it may be time to consider releasing the next version (#421)

**Pull request merged**